core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java

```diff
@@ -161,6 +166,9 @@ public static NodeExplanation calculateNodeExplanation(ShardRouting shard,
         if (node.getId().equals(assignedNodeId)) {
             finalDecision = ClusterAllocationExplanation.FinalDecision.ALREADY_ASSIGNED;
             finalExplanation = "the shard is already assigned to this node";
+        } else if (shard.unassigned() && hasPendingAsyncFetch) {
+            finalExplanation = "the shard state fetch is pending";
```

Comment:
I think this should explain why shard fetching is preventing allocation (otherwise this message will be too cryptic for the user). For primary shards it is to determine which nodes have a non-stale copy of the data. For replica shards it is to determine which nodes have a copy of the data and which copy shares the most data with the primary to speed up recovery.